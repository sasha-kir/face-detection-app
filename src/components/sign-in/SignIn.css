/* adapted from code by marko https://codepen.io/marko-zub/ */

.sign-in-main-div {
	height: 90vh;
}

.sign-in-wrapper {
	width: 450px;
	margin: auto;
	align-self: center;
	background: #212121;
	box-shadow: 0 0 15px rgba( 255, 255, 255, 0.2 );
}

.sign-in-form {
  width: 300px;
  padding: 7px 20px 20px;
  margin: 0 auto;
}

.sign-in-form .form-element {
  margin: 0 0 24px;
}

.sign-in-form .form-element input {
	padding: 14px 16px;
}  

.sign-in-form .block-button {
	padding: 14px 16px;
}

.login-text {
	font-size: 18px;
}

.sign-in-form .block-button:hover, 
.sign-in-form .block-button:focus {
  color: #212121;
  text-shadow: 0.5px 0.5px black;
  background: linear-gradient(90deg, rgba(95,199,230,1) 0%, rgba(234,244,244,1) 100%);
}

.sign-in-form .block-button:focus .cube-inner, 
.sign-in-form .block-button:hover .cube-inner {
  top: 100%;
}

.sign-in-form .block-button:focus .cube-right-inner,
.sign-in-form .block-button:focus .cube-top-inner,
.sign-in-form .block-button:hover .cube-right-inner,
.sign-in-form .block-button:hover .cube-top-inner {
  top: 100%;
}

/* https://ianlunn.github.io/Hover/ */
/*@keyframes hvr-ripple-out {
  100% {
    top: -15px;
    right: -20px;
    bottom: -12px;
    left: -10px;
    opacity: 0;
  }
}

.hvr-ripple-out:before {
  content: '';
  position: absolute;
  border: #e1e1e1 solid 6px;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  animation-duration: 0.2s;
}

.hvr-ripple-out:active:before {
  animation-name: hvr-ripple-out;
}*/