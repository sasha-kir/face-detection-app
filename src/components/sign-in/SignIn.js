import React, { Component } from "react";
import "./SignIn.css";
import { withAlert } from "react-alert";

class SignIn extends Component {
	_isMounted = false;

	constructor(props) {
		super(props);
		this.handleEnter = this.handleEnter.bind(this);
		this.handleSubmit = this.handleSubmit.bind(this);
		this.abortController = new AbortController();
		this.state = {
			username: "",
			password: "",
			dataSubmitted: false,
			dataReceived: null
		};
	}

	componentDidMount() {
		this._isMounted = true;
	}

	componentWillUnmount() {
		this._isMounted = false;
	}

	onUsernameChange = (event) => {
		this.setState({ username: event.target.value });
	}

	onPasswordChange = (event) => {
		this.setState({ password: event.target.value });
	}

	handleEnter(event) {
   		if (event.key === "Enter") {
   			this.handleSubmit();
   		}
   	}
	
	handleSubmit = () => {
		const { username, password } = this.state;
		if (!username || !password) {
			this.props.alert.show("Please fill in all appropriate fields");
		} else {
			this.setState({ dataSubmitted: true });
			fetch(process.env.REACT_APP_SERVER_URL + "/sign-in", {
				method: "post",
				headers: {'Content-Type': 'application/json'},
				body: JSON.stringify({
					username: username,
					password: password
				})
			})
			  .then(response => {
			  	if (response.status === 200) {
			  		return Promise.resolve(response.json())
			  	} else {
			  		var error = new Error(response.statusText || response.status);
      				error.response = response;
      				return Promise.reject(error)
			  	}
			  })
			  .then(user => {
					if (this._isMounted) {
						this.setState({ dataReceived: true });
					}
			      	this.props.loadUser(user);
			        this.props.onRouteChange("home");
			  })
			  .catch(err => {
				if (this._isMounted) {
					this.setState({ dataReceived: false });
				}
			  	this.props.alert.error("Wrong email or password");
			  })
		}
	}

	render() {
		const { onRouteChange } = this.props;
		return (
			<div className="sign-in-main-div center-content">
				<div className="sign-in-wrapper">
					<form className="sign-in-form">
					  <div className="form-element f4 white">
					    <h1>Sign In</h1>
					  </div>

					  <div className="form-element block-cube block-input">
					    <input id="username" placeholder="email" type="text" 
					           onChange={this.onUsernameChange} required />
					    <div className="cube-top"><div className="cube-top-inner"></div></div>
					    <div className="cube-right"><div className="cube-right-inner"></div></div>
					    <div className="cube"><div className="cube-inner"></div></div>
					  </div>

					  <div className="form-element block-cube block-input">
					    <input id="password" placeholder="password" type="password" 
					           onChange={this.onPasswordChange} 
					           onKeyDown={this.handleEnter} required />
					    <div className="cube-top"><div className="cube-top-inner"></div></div>
					    <div className="cube-right"><div className="cube-right-inner"></div></div>
					    <div className="cube"><div className="cube-inner"></div></div>
					  </div>

					  <button className="block-button block-cube grow" 
					  		  type="button" onClick={this.handleSubmit} >
					    <div className="cube-top"><div className="cube-top-inner"></div></div>
					    <div className="cube-right"><div className="cube-right-inner"></div></div>
					    <div className="cube"><div className="cube-inner"></div></div>
					    <div className="login-text button-text">Log In</div>
					  </button>

					  <div id="register-link" className="link dim pointer mt4 mb2 f5 white center-content" 
					       onClick={() => onRouteChange("register")}>
					  	Create an account
					  </div>
					</form>
				</div>
			</div>
		);
	}
}

export default withAlert()(SignIn);
