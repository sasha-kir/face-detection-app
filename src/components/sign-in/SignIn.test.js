import React from 'react';
import {act} from 'react-dom/test-utils';  
import { mount } from 'enzyme';
import nock from 'nock';
import waitUntil from 'async-wait-until';
import { Provider as AlertProvider } from "react-alert";
import AlertTemplate from '../alert-template/react-alert-custom-template';

import SignIn from './SignIn';

const loadUser = jest.fn();
const onRouteChange = jest.fn();

describe('SignIn', () => {
    let component;

    beforeEach(() => {
        component = mount(
            <AlertProvider template={AlertTemplate}>
                <SignIn onRouteChange={onRouteChange} 
                        loadUser={loadUser} />
            </AlertProvider>);
    });

    afterEach(() => {
        component.unmount();
        nock.cleanAll();
    });

    it('renders correctly', () => {
        const wrapper= mount(<SignIn />);
        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
    });
    it('saves username and password to state', () => {
        const form = component.find('SignIn');
        const usernameInput = component.find('input#username').at(0);
        const passInput = component.find('input#password').at(0);

        usernameInput.instance().value = "amy@gmail.com";
        usernameInput.simulate('change');
        expect(form.state('username')).toBe("amy@gmail.com");

        passInput.instance().value = "secret";
        passInput.simulate('change');
        expect(form.state('password')).toBe("secret");
    });
    it('submits on button click', () => {
        const form = component.find('SignIn');
        act(() => {
            form.setState({ username: "amy@gmail.com" });
            form.setState({ password: "secret" });
        });
        form.find('button').simulate('click');
        expect(form.state('dataSubmitted')).toBeTruthy();
    });
    it('submits on Enter in password field', () => {
        const form = component.find('SignIn');
        act(() => {
            form.setState({ username: "amy@gmail.com" });
            form.setState({ password: "secret" });
        });
        form.find('input#password').simulate('keydown', { key: 'Enter' });
        expect(form.state('dataSubmitted')).toBeTruthy();
    });
    it('handles empty fields on submit', () => {
        const form = component.find('SignIn');
        const button = form.find('button');

        button.simulate('click');
        expect(form.state('dataSubmitted')).toBeFalsy();

        act(() => {
            form.setState({ username: "amy@gmail.com" });
            form.setState({ password: "" });
        });

        button.simulate('click');
        expect(form.state('dataSubmitted')).toBeFalsy();

        act(() => {
            form.setState({ username: "" });
            form.setState({ password: "secret" });
        });
        button.simulate('click');
        expect(form.state('dataSubmitted')).toBeFalsy();
    });
    it('sends data to server and handles response 200', async(done) => {
        const mockUser = { id: 1, name: "Amy", email: "amy@gmail.com", entries: 0, joined: "" };

        nock('http://localhost:3001')
            //.log(console.log)
            .persist()
            .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
            .post('/sign-in', {username: 'amy@gmail.com', password: 'secret'})
            .reply(200, mockUser);
        
        const form = component.find('SignIn');
        act(() => {
            form.setState({ username: "amy@gmail.com" });
            form.setState({ password: "secret" });
        });
        form.find('button').simulate('click');
        await act(async() => {
            await waitUntil(() => form.state('dataReceived') !== null);
        });
        expect(form.state('dataReceived')).toBeTruthy();
        expect(loadUser).toHaveBeenCalledWith(mockUser);
        expect(onRouteChange).toHaveBeenCalledWith('home');
        done();
    });
    it('handles error response from server', async(done) => {
        nock('http://localhost:3001')
            //.log(console.log)
            .persist()
            .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
            .post('/sign-in', {username: 'amy@gmail.com', password: 'secret'})
            .reply(400);

        const form = component.find('SignIn');
        act(() => {
            form.setState({ username: "amy@gmail.com" });
            form.setState({ password: "secret" });
        });
        form.find('button').simulate('click');
        await act(async() => {
            await waitUntil(() => form.state('dataReceived') !== null);
        });
        expect(form.state('dataReceived')).toBe(false);
        done();
    });
});