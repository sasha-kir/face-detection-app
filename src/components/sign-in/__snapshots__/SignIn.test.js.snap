// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn renders correctly 1`] = `
<ForwardRef(WithAlert(SignIn))>
  <SignIn>
    <div
      className="sign-in-main-div center-content"
    >
      <div
        className="sign-in-wrapper"
      >
        <form
          className="sign-in-form"
        >
          <div
            className="form-element f4 white"
          >
            <h1>
              Sign In
            </h1>
          </div>
          <div
            className="form-element block-cube block-input"
          >
            <input
              id="username"
              onChange={[Function]}
              placeholder="username"
              required={true}
              type="text"
            />
            <div
              className="cube-top"
            >
              <div
                className="cube-top-inner"
              />
            </div>
            <div
              className="cube-right"
            >
              <div
                className="cube-right-inner"
              />
            </div>
            <div
              className="cube"
            >
              <div
                className="cube-inner"
              />
            </div>
          </div>
          <div
            className="form-element block-cube block-input"
          >
            <input
              id="password"
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="password"
              required={true}
              type="password"
            />
            <div
              className="cube-top"
            >
              <div
                className="cube-top-inner"
              />
            </div>
            <div
              className="cube-right"
            >
              <div
                className="cube-right-inner"
              />
            </div>
            <div
              className="cube"
            >
              <div
                className="cube-inner"
              />
            </div>
          </div>
          <button
            className="block-button block-cube grow"
            onClick={[Function]}
            type="button"
          >
            <div
              className="cube-top"
            >
              <div
                className="cube-top-inner"
              />
            </div>
            <div
              className="cube-right"
            >
              <div
                className="cube-right-inner"
              />
            </div>
            <div
              className="cube"
            >
              <div
                className="cube-inner"
              />
            </div>
            <div
              className="login-text button-text"
            >
              Log In
            </div>
          </button>
          <div
            className="link dim pointer mt4 mb2 f5 white center-content"
            id="register-link"
            onClick={[Function]}
          >
            Create an account
          </div>
        </form>
      </div>
    </div>
  </SignIn>
</ForwardRef(WithAlert(SignIn))>
`;
